ADR for Localization (#641)
bitwarden/contributing-docs: docs/architecture/adr/0027-angular-i18n-migration.md

---
adr: "0027"
status: Proposed
date: 2025-08-01
tags: [clients, angular, i18n]
---

# 0027 - Improved Localization

<AdrTable frontMatter={frontMatter}></AdrTable>

## Context and Problem Statement

We currently use a homegrown localization system based on the

> https://developer.chrome.com/docs/extensions/reference/api/i18n api. This has served us reasonably

Reviewer: 🎨 Make into a prettier link and not just show the URL? There are a few of these on this page.

> well but comes with many limitations which we've had to workaround over the years.

Reviewer: ```suggestion
well but comes with many limitations which we've had to work around over the years.
```

> Most notable it lacks any native capability for interpolation which is useful for adding markdown

Reviewer: ```suggestion
Most notable it lacks any native capability for interpolation which is useful for adding Markdown
```

Author: This was supposed to be `markup` 🤦

> Most notable it lacks any native capability for interpolation which is useful for adding markdown

Reviewer: ```suggestion
Most notably it lacks any native capability for interpolation which is useful for adding markdown
```
🎨 I wouldn't make this a new paragraph since you're continuing the topic with "most notably".

such as marking words as italic or bold or inserting links into strings. In most scenarios we've
handled this by splitting the sentence into multiple localized strings but that is not always
optimal in all locales.

Another limitation is pluralization where we currently can't distinguish phrases that are singular
or plural which we've often mitigated by adding (s) at the end which again provides a sub-optimal

> experience as different languages handles names, numbers and dates differently.

Reviewer: ```suggestion
experience as different languages handles names, numbers, and dates differently.
```

In order to better handle more complex localization scenarios we'd like to explore changing the
underlying localization engine.

## Considered Options

- **Keep chrome.i18n** - Continue using our current homegrown localization system
- **Continue building our own** - Expand our custom localization framework
- **Angular Localization (@angular/localize)** - Use Angular's built-in i18n framework
- **ngx-translate** - Use the popular Angular translation library
- **`Transloco`** - Use the modern Angular translation library
- **angular-i18next** - Use i18next with Angular integration

### Keep `chrome.i18n`

> **Pros:**

Reviewer: 🎨 I would prefer using `####` for these and dropping the colon.

Author: I dropped the colon but didn't make them titles since they would show up in the sidebar which would just be noisy.

- We already use it and have existing infrastructure
- No migration effort required

**Cons:**

- Lacks native interpolation capabilities for HTML
- No pluralization support

### Continue building our own

**Pros:**

- Complete control over features and implementation
- Can be tailored exactly to our needs

**Cons:**

- Development and maintenance overhead
- Localization is a solved problem with standardized formats

### Angular Localization (`@angular/localize`)

**Pros:**

- Built-in framework for Angular with long-term support
- Expected to be supported for the lifetime of Angular
- Seamless integration in Angular environments
- Advanced capabilities like pluralization and interpolation

> - Uses standardized ICU MessageFormat

Reviewer: ```suggestion
- Uses standardized ICU `MessageFormat`
```
⛏️ A bunch of these technical references could become code.

  }
}
```

#### Content Scripts

Outside Angular environments, we're limited to the `$localize` tag functionality, which provides
basic string interpolation equivalent to our current system. For advanced features, we can combine
with post-processing libraries like `intl-messageformat`:

```typescript
formatIcu($localize`{count, plural, =0 {none} =1 {one} other {more}}`);
```

### Positive Consequences

- Native support for pluralization and interpolation
- Better localization experience across all languages
- Long-term support as part of Angular core
- Reduced need for workarounds and string splitting

### Negative Consequences

- Migration effort required from current system
- Limited functionality in non-Angular environments
- May require additional libraries for content scripts
- Learning curve for development team

### Implementation Plan

Replacing a localization system is difficult since it's used in virtually every component and in
several background services and content scripts. We therefore need to maintain both systems in
parallel and ideally offer automated migrations wherever possible.

> 1. **Proof of concept evaluation**: https://github.com/bitwarden/clients/pull/13737

Reviewer: 🎨 I don't think we need to use boldface in any of these.
